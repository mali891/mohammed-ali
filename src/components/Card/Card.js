import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { HeadingBlock, Link } from '../'
import { styles } from './Card.styles'

const Card = ({ children, className, condensed, link, path, subtitle, title, ...other }) => {
  const dynamicClassNames = classNames([className], {
    'c-card': true,
    'u-bg--tertiary': true,
    'c-card--condensed': condensed,
    'c-card--link': link,
    'c-card-fullheight': link
  })

  const renderCard = () => (
    <div css={link ? null : styles} className={link ? 'c-card--fullheight' : dynamicClassNames} {...other}>
      <div className="c-card__heading">
        <HeadingBlock type="span" size="sm" noMargin>
          {title}
        </HeadingBlock>
        {subtitle && <span className="c-card__subtitle">{subtitle}</span>}
      </div>

      <div className="c-card__content">{children}</div>
    </div>
  )

  return (
    <Fragment>
      {link ? (
        <Link css={styles} className={dynamicClassNames} to={path} noHover>
          {renderCard()}
        </Link>
      ) : (
        <Fragment>{renderCard()}</Fragment>
      )}
    </Fragment>
  )
}

Card.defaultProps = {
  condensed: false,
  link: false
}

Card.propTypes = {
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  className: PropTypes.string,
  condensed: PropTypes.bool,
  link: PropTypes.bool,
  path: PropTypes.string,
  subtitle: PropTypes.string,
  title: PropTypes.string.isRequired
}

export default Card
