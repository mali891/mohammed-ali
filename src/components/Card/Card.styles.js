import { css } from '@emotion/core'
import { COLOURS, FONT_SIZES, EASINGS, SPACING, BORDER_RADIUS, LINE_HEIGHT, MEDIA_QUERIES } from '../../styles/settings'

export const styles = css`
  width: 100%;
  min-height: 30rem;
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  margin: 0 0 ${SPACING.MOBILE.SM} 0;
  transition: all ${EASINGS.QUICK};
  border-radius: ${BORDER_RADIUS.LG} ${BORDER_RADIUS.LG} 0 0;
  padding: ${SPACING.MOBILE.MD};

  &:hover {
    &.c-card--link {
      .c-card__border-bottom {
        &:after {
          width: 100%;
        }
      }
    }
  }

  aside {
    height: 100%;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }

  &.c-card--condensed {
    min-height: 20rem;

    .c-card {
      &__heading {
        margin-bottom: ${SPACING.MOBILE.MD};
      }
    }
  }

  .c-card__heading {
    position: relative;
    word-wrap: break-word;
    hyphens: manual;

    &:after {
      content: '';
      width: 100%;
      position: absolute;
      left: 0;
      bottom: -${SPACING.MD};
      right: 0;
      height: 0.2rem;
      width: 10rem;
      transition: all ${EASINGS.DEFAULT};
      background-image: linear-gradient(to right, ${COLOURS.GRADIENT_THREE.FIRST}, ${COLOURS.GRADIENT_THREE.SECOND});
      border-radius: ${BORDER_RADIUS.LG};
    }
  }

  .c-card__subtitle {
    font-size: ${FONT_SIZES.MOBILE.XS};
    line-height: ${LINE_HEIGHT.MOBILE.XS};
  }

  .c-card__content {
    width: 100%;
    display: flex;
    justify-content: space-between;
    font-size: ${FONT_SIZES.MOBILE.XS};
    line-height: ${LINE_HEIGHT.MOBILE.XS};
    margin-top: ${SPACING.MOBILE.XL};

    p {
      font-size: ${FONT_SIZES.MOBILE.SM};
      line-height: ${LINE_HEIGHT.MOBILE.SM};
    }
  }

  .c-card__border-bottom {
    width: 100%;
    position: absolute;
    left: 0;
    bottom: 0;
    right: 0;
    height: 0.5rem;
    background-color: var(--colour-secondary);
    border-radius: 0 0 ${BORDER_RADIUS.LG} ${BORDER_RADIUS.LG};

    &:after {
      content: '';
      width: 0;
      position: absolute;
      left: 0;
      bottom: 0;
      right: 0;
      height: 0.2rem;
      width: 0;
      height: 0.5rem;
      transition: all ${EASINGS.DEFAULT};
      background-image: linear-gradient(to right, ${COLOURS.GRADIENT_THREE.FIRST}, ${COLOURS.GRADIENT_THREE.SECOND});
      border-radius: 0 0 ${BORDER_RADIUS.LG} ${BORDER_RADIUS.LG};
    }
  }

  @media (min-width: ${MEDIA_QUERIES.TABLET}) {
    width: 48%;
    padding: ${SPACING.TABLET.MD};

    &:nth-of-type(odd) {
      margin-right: 1%;
    }

    &:nth-of-type(even) {
      margin-left: 1%;
    }
  }

  @media (min-width: ${MEDIA_QUERIES.NOTEBOOK}) {
    padding: ${SPACING.TABLET.MD};

    .c-card__border-bottom {
      height: ${SPACING.MOBILE.XS};

      &:after {
        height: ${SPACING.MOBILE.XS};
      }
    }
  }

  @media (min-width: ${MEDIA_QUERIES.DESKTOP}) {
    width: 40rem;
    max-width: 20%;
    min-height: 50rem;
    margin: 0 1%;
    padding: ${SPACING.DESKTOP.MD};

    &.c-card--condensed {
      min-height: 40rem;

      .c-card {
        &__heading {
          margin-bottom: ${SPACING.DESKTOP.MD};
        }
      }
    }

    .c-card__heading {
      margin-bottom: ${SPACING.DESKTOP.XL};

      &:after {
        bottom: -${SPACING.DESKTOP.MD};
        width: 10rem;
      }
    }

    .c-card__subtitle {
      font-size: ${FONT_SIZES.DESKTOP.XS};
      line-height: ${LINE_HEIGHT.DESKTOP.XS};
    }

    .c-card__content {
      font-size: ${FONT_SIZES.DESKTOP.XS};
    }

    .c-card__border-bottom {
      height: ${SPACING.DESKTOP.XS};

      &:after {
        height: ${SPACING.DESKTOP.XS};
      }
    }
  }
`
