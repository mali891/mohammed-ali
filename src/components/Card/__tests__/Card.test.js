import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Card from '../Card'

const defaultProps = {
  title: 'Dummy title'
}

describe('Card', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Card {...defaultProps} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly given a fullWidth prop', () => {
    const { asFragment, container } = render(<Card {...defaultProps} fullWidth />)

    expect(container.querySelector('.c-card--fullwidth')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders a link when given a link prop', () => {
    const { asFragment, container } = render(<Card {...defaultProps} link />)

    expect(container.querySelector('.c-link')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders a subtitle when given a subtitle prop', () => {
    const { asFragment, container } = render(<Card {...defaultProps} subtitle="Dummy subtitle" />)

    expect(container.querySelector('.c-card__subtitle')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })
})
