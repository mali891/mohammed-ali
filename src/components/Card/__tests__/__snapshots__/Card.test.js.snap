// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders a link when given a link prop 1`] = `
<DocumentFragment>
  <a
    class="c-card u-bg--tertiary c-card--link c-card-fullheight c-link u-clr--secondary c-link--no-hover"
    css="[object Object]"
  >
    <div
      class="c-card--fullheight"
    >
      <div
        class="c-card__heading"
      >
        <span
          class="c-heading c-heading--sm c-heading--no-margin"
          css="[object Object]"
        >
          Dummy title
        </span>
      </div>
      <div
        class="c-card__content"
      />
    </div>
  </a>
</DocumentFragment>
`;

exports[`Card renders a subtitle when given a subtitle prop 1`] = `
<DocumentFragment>
  <div
    class="c-card u-bg--tertiary"
    css="[object Object]"
  >
    <div
      class="c-card__heading"
    >
      <span
        class="c-heading c-heading--sm c-heading--no-margin"
        css="[object Object]"
      >
        Dummy title
      </span>
      <span
        class="c-card__subtitle"
      >
        Dummy subtitle
      </span>
    </div>
    <div
      class="c-card__content"
    />
  </div>
</DocumentFragment>
`;

exports[`Card renders correctly 1`] = `
<DocumentFragment>
  <div
    class="c-card u-bg--tertiary"
    css="[object Object]"
  >
    <div
      class="c-card__heading"
    >
      <span
        class="c-heading c-heading--sm c-heading--no-margin"
        css="[object Object]"
      >
        Dummy title
      </span>
    </div>
    <div
      class="c-card__content"
    />
  </div>
</DocumentFragment>
`;

exports[`Card renders correctly given a fullWidth prop 1`] = `
<DocumentFragment>
  <div
    class="c-card u-bg--tertiary c-card--fullwidth"
    css="[object Object]"
  >
    <div
      class="c-card__heading"
    >
      <span
        class="c-heading c-heading--sm c-heading--no-margin"
        css="[object Object]"
      >
        Dummy title
      </span>
    </div>
    <div
      class="c-card__content"
    />
  </div>
</DocumentFragment>
`;
